packages/@uppy/aws-s3-multipart/src/index.js

```diff
@@ -17,11 +18,189 @@ function assertServerError (res) {
+    }
+
+    const isMutatingRetryDelay = 'retryDelays' in options
+    if (isMutatingRetryDelay) {
+      this.#retryDelayIterator = options.retryDelays?.values()
+    }
+    const isMutatingUploadPartBytes = 'uploadPartBytes' in options
+    if (isMutatingUploadPartBytes) {
+      this.#rawUploadPartBytes = options.uploadPartBytes
+    }
+    if (isMutatingUploadPartBytes || isMutatingRetryDelay) {
+      const retryDelaysIterator = this.#retryDelayIterator
+      // Requests to Amazon server are the highest priority because we want the upload to
+      // start as soon we got the signature to limit the risk of the signature expiring.
+      const uploadPartBytes = requests.wrapPromiseFunction(this.#rawUploadPartBytes, { priority:Infinity })
+      // TODO: we currently only retry the PUT but we also want to retry the signing request.
+      // TODO: this retry logic is taken out of Tus. We should have a centralized place for retrying,
+      // perhaps the rate limited queue, and dedupe all plugins with that.
+      this.#uploadPartBytes = async (...args) => {
+        for (;;) {
+          try {
+            return await uploadPartBytes(...args)
+          } catch (err) {
+            const status = err?.source?.status
+            if (status == null) {
+              throw err
+            } else if (status === 429) {
+              // HTTP 429 Too Many Requests => to avoid the whole download to fail, pause all requests.
+              if (!requests.isPaused) {
+                const next = retryDelaysIterator?.next()
+                if (next == null || next.done) {
+                  throw err
+                }
+                requests.rateLimit(next.value)
+              }
+            } else if (status > 400 && status < 500 && status !== 409) {
+              // HTTP 4xx, the server won't send anything, it's doesn't make sense to retry
+              throw err
+            } else if (typeof navigator !== 'undefined' && navigator.onLine === false) {
+              // The navigator is offline, let's wait for it to come back online.
+              if (!requests.isPaused) {
+                requests.pause()
+                window.addEventListener('online', () => {
+                  requests.resume()
+                }, { once: true })
+              }
+            } else {
+              // Other error code means the request can be retried later.
+              const next = retryDelaysIterator?.next()
+              if (next == null || next.done) {
+                throw err
+              }
+              await new Promise(resolve => setTimeout(resolve, next.value))
+            }
+          }
+        }
+      }
+    }
+  }
+
+  async getUploadId (file, signal) {
+    const cachedResult = this.#cache.get(file.data)
+    if (cachedResult != null) {
+      return cachedResult
+    }
+
+    const promise = this.#createMultipartUpload(file, signal).then(async (result) => {
```

Comment:
do we need to catch this error?

Reply:
Surtout pas ! If this fails, we want the error to bubble up.